pdns/pdns_recursor.cc

```diff
@@ -396,114 +396,113 @@ static bool isHandlerThread()
+LWResult::Result arecvtcp(PacketBuffer& data, const size_t len, shared_ptr<TCPIOHandler>& handler, const bool incompleteOkay)
 {
-  data.clear();
-  PacketID pident;
-  pident.tcpsock=sock->getHandle();
-  pident.inNeeded=len;
-  pident.inIncompleteOkay=incompleteOkay;
-  t_fdm->addReadFD(sock->getHandle(), handleTCPClientReadable, pident);
+  TCPLOG("arecvtcp called " << len << ' ' << data.size() << endl);
+  data.resize(len);
 
-  int ret = MT->waitEvent(pident, &data, g_networkTimeoutMsec);
-  if (ret == 0) {
-    t_fdm->removeReadFD(sock->getHandle());
-    return LWResult::Result::Timeout;
-  }
-  else if (ret == -1) {
-    t_fdm->removeWriteFD(sock->getHandle());
-    return LWResult::Result::PermanentError;
+  // We might have data already available from the TLS layer, try to get that into the buffer
+  size_t pos = 0;
+  IOState state;
+  try {
+    TCPLOG("calling tryRead() " << len << endl);
+    state = handler->tryRead(data, pos, len);
+    TCPLOG("arcvtcp tryRead() returned " << int(state) << ' ' << pos << '/' << len << endl);
+    switch (state) {
+    case IOState::Done:
+    case IOState::NeedRead:
+      if (pos == len || (incompleteOkay && pos > 0)) {
+        data.resize(pos);
+        TCPLOG("acecvtcp success A" << endl);
+        return LWResult::Result::Success;
+      }
+      break;
+    case IOState::NeedWrite:
+      break;
+    }
   }
-  else if (data.empty()) {// error, EOF or other
+  catch (const std::exception& e) {
+    TCPLOG("tryRead() exception..." << e.what() << endl);
     return LWResult::Result::PermanentError;
   }
 
-  return LWResult::Result::Success;
-}
-
-static void TCPIOHandlerReadable(int fd, FDMultiplexer::funcparam_t& var);
+  auto pident = std::make_shared<PacketID>();
+  pident->tcphandler = handler;
+  pident->tcpsock = handler->getDescriptor();
+  pident->inNeeded = len;
+  pident->inIncompleteOkay = incompleteOkay;
+  pident->highState = TCPAction::DoingRead;
+  pident->lowState = state;
 
-LWResult::Result arecvtcp(PacketBuffer& data, const size_t len, TCPIOHandler& handler, const bool incompleteOkay)
-{
   data.clear();
 
-  PacketID pident;
-  pident.tcphandler = &handler;
-  pident.tcpsock = handler.getDescriptor();
-  pident.inNeeded = len;
-  pident.inIncompleteOkay = incompleteOkay;
-  t_fdm->addReadFD(handler.getDescriptor(), TCPIOHandlerReadable, pident);
+  if (state == IOState::NeedWrite) {
+    t_fdm->addWriteFD(handler->getDescriptor(), TCPIOHandlerIO, pident);
+  } else {
+    t_fdm->addReadFD(handler->getDescriptor(), TCPIOHandlerIO, pident);
+  }
 
-  int ret = MT->waitEvent(pident, &data, g_networkTimeoutMsec);
+  int ret = MT->waitEvent(*pident, &data, g_networkTimeoutMsec);
+  TCPLOG("arecvtcp " << ret << ' ' << data.size() << ' ' );
   if (ret == 0) {
-    t_fdm->removeReadFD(handler.getDescriptor());
+    TCPLOG("timeout" << endl);
+    t_fdm->removeReadFD(handler->getDescriptor());
```

Comment:
What if we were waiting for the ability to write?

Reply:
Handled, including the other cases.

Reply:
Are you sure we are not going to try to remove the FD from the "waiting for read" set of the multiplexer if a timeout occurs while we are waiting for the FD to become writable?

Reply:
Riught, I'm going to fix that by calling `TCPIOHandlerStateChange`